File: baekjoon/disjointset/1717_집합의표현.java
import java.io.*;
import java.util.*;

public class Main {

    static int[] parent;

    static int find(int a){

        if(a != parent[a]){
            parent[a] = find(parent[a]);
        }
        return parent[a];
    }
    
    static void union(int a, int b){
        a = find(a);
        b = find(b);

        if(a > b){
            parent[b] = a;
        }
        else{
            parent[a] = b;
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        StringBuilder sb = new StringBuilder();

        parent = new int[n+1];

        for(int i = 0; i < n + 1; i++){
            parent[i] = i;
        }

        for(int i = 0; i < m ; i++){
            st = new StringTokenizer(br.readLine(), " ");

            int op = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

              if (op == 1){
                if(find(a) == find(b)){
                    sb.append("YES\n");
                }
                else{
                    sb.append("NO\n");
                }
            }
            else{
                union(a, b);
            }
        }
        System.out.println(sb.toString());
    }
}

File: baekjoon/disjointset/1717_집합의표현.py
from sys import (
    stdin, 
    setrecursionlimit
    )

setrecursionlimit(100000)
parent = []

def uni(a, b):
    a = find(a)
    b = find(b)

    if a < b:
        parent[b] = a
    else :
        parent[a] = b

def find(a):
    if a != parent[a]:
        parent[a] = find(parent[a])
    return parent[a]

def solution():
    global parent
    n, m = map(int, stdin.readline().split())

    parent = [i for i in range(n+1)]

    for _ in range(m):
        op, a, b = map(int, stdin.readline().split())

        if op: # op == 1, 두 원소가 같은 집한에 있는지 확인
            if find(a) == find(b):
                print("YES")
            else:
                print("NO")
        else: # op == 0, 합 집합
            uni(a, b)

solution()